src/UCTNodeRoot.cpp

```diff
@@ -201,6 +400,146 @@ void UCTNode::prepare_root_node(Network & network, int color,
     // This also removes a lot of special cases.
     kill_superkos(root_state);
 
+    if (had_children) {
+        root_eval = get_eval(color);
+    }
+    else {
+        if (root_state.eval_invalid()) {
+            GameState tmpstate = root_state;
+            tmpstate.play_move(get_first_child()->get_move());
+            root_eval = 1.0f - white_net_eval(network, tmpstate);
+        }
+        update(root_eval);
+        root_eval = (color == FastBoard::BLACK ? root_eval : 1.0f - root_eval);
+    }
+
+    std::array<std::vector<std::shared_ptr<Sym_State>>, 2> ss;
```

Comment:
This appears to collect the average eval during the search somehow. Not clear why it's so complicated, or why it's tracking symmetries.

Reply:
> Instead of collecting the 200 positions during the search, and then taking a 4% sample afterwards, what about making the komi adjustment loop perform a minimal search (4% of 200 positions is 8 nodes) with the adjusted parameters? The main disadvantage is that the positions would be closer together.

> Or go back to only adjusting at the root (what you seem to have initially had) for a first merge attempt, and then try to clean up the position collection afterwards? It seems to be almost half of this patch, and it's some of the messiest code.

The code can be made less complicated by removing some inferior/abandoned ideas/features. The winrate can differ a lot with different symmetries, but it's still reasonable to expect monotonicity (of winrate w.r.t. komi) for each fixed symmetry. For a set of (position, symmetry) pairs, mean_white_eval returns a fixed value for each komi. Under the assumption that the returned value is monotonic w.r.t. komi, binary search will find a komi that attains the target winrate. Due to symmetry randomness and resulting node expansion randomness, the returned value of a minimal search isn't deterministic, let alone monotonic; and you'll need to run the net for all nodes, not just 4% of them.

Originally (in komi-v0.3x) I just used the 8 symmetries of the root node to adjust komi, but it's observed by @kuba97531 for example in a 4H game against an EGF 7d that if the winrate drops after some search, crazy moves will appear, and my idea to solve the problem is to collect positions deeper into the search for komi adjustment.